src/qfusion.cpp

```diff
@@ -0,0 +1,541 @@
+        rand_generator->seed(std::time(0));
+    } else {
+        rand_generator = rgp;
+    }
+
+    qReg = CreateQuantumInterface(engineType, qBitCount, initState, rand_generator);
+}
+
+void QFusion::ApplySingleBit(const complex* mtrx, bool doCalcNorm, bitLenInt qubitIndex)
+{
+    if (qubitCount < MIN_FUSION_BITS) {
+        qReg->ApplySingleBit(mtrx, doCalcNorm, qubitIndex);
+        return;
+    }
+
+    std::shared_ptr<complex[4]> outBuffer(new complex[4]);
+    if (bitBuffers[qubitIndex]) {
+        std::shared_ptr<complex[4]> inBuffer = bitBuffers[qubitIndex];
+        std::vector<std::future<void>> futures(4);
+
+        futures[0] = std::async(std::launch::async, [&]() {
+            outBuffer[0] = (mtrx[0] * inBuffer[0]) + (mtrx[1] * inBuffer[2]);
+            if (norm(outBuffer[0]) < min_norm) {
+                outBuffer[0] = complex(ZERO_R1, ZERO_R1);
+            }
+        });
+        futures[1] = std::async(std::launch::async, [&]() {
+            outBuffer[1] = (mtrx[0] * inBuffer[1]) + (mtrx[1] * inBuffer[3]);
+            if (norm(outBuffer[1]) < min_norm) {
+                outBuffer[1] = complex(ZERO_R1, ZERO_R1);
+            }
+        });
+        futures[2] = std::async(std::launch::async, [&]() {
+            outBuffer[2] = (mtrx[2] * inBuffer[0]) + (mtrx[3] * inBuffer[2]);
+            if (norm(outBuffer[2]) < min_norm) {
+                outBuffer[2] = complex(ZERO_R1, ZERO_R1);
+            }
+        });
+        futures[3] = std::async(std::launch::async, [&]() {
+            outBuffer[3] = (mtrx[2] * inBuffer[1]) + (mtrx[3] * inBuffer[3]);
+            if (norm(outBuffer[3]) < min_norm) {
+                outBuffer[3] = complex(ZERO_R1, ZERO_R1);
+            }
+        });
+
+        for (int i = 0; i < 4; i++) {
+            futures[i].get();
+        }
+    } else {
+        std::copy(mtrx, mtrx + 4, outBuffer.get());
+    }
+
+    bitBuffers[qubitIndex] = outBuffer;
+}
+
+void QFusion::SetQuantumState(complex* inputState)
+{
+    DiscardAll();
+    qReg->SetQuantumState(inputState);
+}
+
+void QFusion::GetQuantumState(complex* outputState)
+{
+    FlushAll();
+    qReg->GetQuantumState(outputState);
+}
+
+complex QFusion::GetAmplitude(bitCapInt perm)
+{
+    FlushAll();
+    return qReg->GetAmplitude(perm);
+}
+
+void QFusion::SetPermutation(bitCapInt perm)
+{
+    DiscardAll();
+    qReg->SetPermutation(perm);
+}
+
+void QFusion::SetReg(bitLenInt start, bitLenInt length, bitCapInt value)
+{
+    DiscardReg(start, length);
+    qReg->SetReg(start, length, value);
+}
+
+void QFusion::SetBit(bitLenInt qubitIndex, bool value)
+{
+    DiscardBit(qubitIndex);
+    qReg->SetBit(qubitIndex, value);
+}
+
+bitLenInt QFusion::Cohere(QFusionPtr toCopy)
+{
+    FlushAll();
+    toCopy->FlushAll();
+    bitLenInt toRet = qReg->Cohere(toCopy->qReg);
+    SetQubitCount(qReg->GetQubitCount());
+    return toRet;
+}
+
+void QFusion::Decohere(bitLenInt start, bitLenInt length, QFusionPtr dest)
```

Comment:
Add blank lines between operationally-distinct sections for added clarity, and include commentary describing the steps taken.

Reply:
See my next commit. I've documented the purpose and method.